// @tarojs/vite-runner@4.0.6/dist/utils/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HARMONY_SCOPES = exports.H5_EXCLUDE_POSTCSS_PLUGIN_NAME = exports.MINI_EXCLUDE_POSTCSS_PLUGIN_NAME = exports.DEFAULT_TERSER_OPTIONS = exports.PAGENAME_QUERY = exports.ENTRY_QUERY = exports.customWrapperName = exports.baseCompName = exports.baseTemplateName = exports.backSlashRegEx = exports.quoteNewlineRegEx = exports.needsEscapeRegEx = void 0;
exports.needsEscapeRegEx = /[\n\r'\\\u2028\u2029]/;
exports.quoteNewlineRegEx = /([\n\r'\u2028\u2029])/g;
exports.backSlashRegEx = /\\/g;
exports.baseTemplateName = 'base';
exports.baseCompName = 'comp';
exports.customWrapperName = 'custom-wrapper';
exports.ENTRY_QUERY = 'taro-entry-loader';
exports.PAGENAME_QUERY = 'taro-pagename';
exports.DEFAULT_TERSER_OPTIONS = {
    parse: {
        ecma: 8,
    },
    compress: {
        ecma: 5,
        warnings: false,
        arrows: false,
        collapse_vars: false,
        comparisons: false,
        computed_props: false,
        hoist_funs: false,
        hoist_props: false,
        hoist_vars: false,
        inline: false,
        loops: false,
        negate_iife: false,
        properties: false,
        reduce_funcs: false,
        reduce_vars: false,
        switches: false,
        toplevel: false,
        typeofs: false,
        booleans: true,
        if_return: true,
        sequences: true,
        unused: true,
        conditionals: true,
        dead_code: true,
        evaluate: true,
    },
    output: {
        ecma: 5,
        comments: false,
        ascii_only: true,
    },
};
exports.MINI_EXCLUDE_POSTCSS_PLUGIN_NAME = ['cssModules'];
exports.H5_EXCLUDE_POSTCSS_PLUGIN_NAME = ['cssModules', 'url'];
exports.HARMONY_SCOPES = [/^@system\./, /^@ohos\./, /^@hmscore\//, /^@jd-oh\//];
