// @tarojs/components@4.0.7/lib/solid/component-lib/index.js
import { Fragment } from 'solid-js/h/jsx-runtime';
import '../solid-component-lib/index.js';
import { createSolidComponent } from '../solid-component-lib/createComponent.js';

// 视图容器
const CoverImage = /* @__PURE__ */ createSolidComponent('taro-cover-image-core');
const CoverView = /* @__PURE__ */ createSolidComponent('taro-cover-view-core');
const MatchMedia = /* @__PURE__ */ createSolidComponent('taro-match-media-core');
const MovableArea = /* @__PURE__ */ createSolidComponent('taro-movable-area-core');
const MovableView = /* @__PURE__ */ createSolidComponent('taro-movable-view-core');
const PageContainer = /* @__PURE__ */ createSolidComponent('taro-page-container-core');
const RootPortal = /* @__PURE__ */ createSolidComponent('taro-root-portal-core');
const ScrollView = /* @__PURE__ */ createSolidComponent('taro-scroll-view-core');
const ShareElement = /* @__PURE__ */ createSolidComponent('taro-share-element-core');
const Swiper = /* @__PURE__ */ createSolidComponent('taro-swiper-core');
const SwiperItem = /* @__PURE__ */ createSolidComponent('taro-swiper-item-core');
const View = /* @__PURE__ */ createSolidComponent('taro-view-core');
// 基础内容
const Icon = /* @__PURE__ */ createSolidComponent('taro-icon-core');
const Progress = /* @__PURE__ */ createSolidComponent('taro-progress-core');
const RichText = /* @__PURE__ */ createSolidComponent('taro-rich-text-core');
const Text = /* @__PURE__ */ createSolidComponent('taro-text-core');
// 表单组件
const Button = /* @__PURE__ */ createSolidComponent('taro-button-core');
const Checkbox = /* @__PURE__ */ createSolidComponent('taro-checkbox-core');
const CheckboxGroup = /* @__PURE__ */ createSolidComponent('taro-checkbox-group-core');
const Editor = /* @__PURE__ */ createSolidComponent('taro-editor-core');
const Form = /* @__PURE__ */ createSolidComponent('taro-form-core');
const Input = /* @__PURE__ */ createSolidComponent('taro-input-core');
const KeyboardAccessory = /* @__PURE__ */ createSolidComponent('taro-keyboard-accessory-core');
const Label = /* @__PURE__ */ createSolidComponent('taro-label-core');
const Picker = /* @__PURE__ */ createSolidComponent('taro-picker-core');
const PickerView = /* @__PURE__ */ createSolidComponent('taro-picker-view-core');
const PickerViewColumn = /* @__PURE__ */ createSolidComponent('taro-picker-view-column-core');
const Radio = /* @__PURE__ */ createSolidComponent('taro-radio-core');
const RadioGroup = /* @__PURE__ */ createSolidComponent('taro-radio-group-core');
const Slider = /* @__PURE__ */ createSolidComponent('taro-slider-core');
const Switch = /* @__PURE__ */ createSolidComponent('taro-switch-core');
const Textarea = /* @__PURE__ */ createSolidComponent('taro-textarea-core');
// 导航
const FunctionalPageNavigator = /* @__PURE__ */ createSolidComponent('taro-functional-page-navigator-core');
const Navigator = /* @__PURE__ */ createSolidComponent('taro-navigator-core');
const NavigationBar = /* @__PURE__ */ createSolidComponent('taro-navigation-bar-core');
// 媒体组件
const Audio = /* @__PURE__ */ createSolidComponent('taro-audio-core');
const Camera = /* @__PURE__ */ createSolidComponent('taro-camera-core');
const Image = /* @__PURE__ */ createSolidComponent('taro-image-core');
const LivePlayer = /* @__PURE__ */ createSolidComponent('taro-live-player-core');
const LivePusher = /* @__PURE__ */ createSolidComponent('taro-live-pusher-core');
const Video = /* @__PURE__ */ createSolidComponent('taro-video-core');
const VoipRoom = /* @__PURE__ */ createSolidComponent('taro-voip-room-core');
// 地图
const Map = /* @__PURE__ */ createSolidComponent('taro-map-core');
// 画布
const Canvas = /* @__PURE__ */ createSolidComponent('taro-canvas-core');
// 开放能力
const Ad = /* @__PURE__ */ createSolidComponent('taro-ad-core');
const AdCustom = /* @__PURE__ */ createSolidComponent('taro-ad-custom-core');
const OfficialAccount = /* @__PURE__ */ createSolidComponent('taro-official-account-core');
const OpenData = /* @__PURE__ */ createSolidComponent('taro-open-data-core');
const WebView = /* @__PURE__ */ createSolidComponent('taro-web-view-core');
// 页面属性配置节点
const PageMeta = /* @__PURE__ */ createSolidComponent('taro-page-meta-core');
// 其他
const Block = Fragment;
const CustomWrapper = /* @__PURE__ */ createSolidComponent('taro-custom-wrapper-core');

export { Ad, AdCustom, Audio, Block, Button, Camera, Canvas, Checkbox, CheckboxGroup, CoverImage, CoverView, CustomWrapper, Editor, Form, FunctionalPageNavigator, Icon, Image, Input, KeyboardAccessory, Label, LivePlayer, LivePusher, Map, MatchMedia, MovableArea, MovableView, NavigationBar, Navigator, OfficialAccount, OpenData, PageContainer, PageMeta, Picker, PickerView, PickerViewColumn, Progress, Radio, RadioGroup, RichText, RootPortal, ScrollView, ShareElement, Slider, Swiper, SwiperItem, Switch, Text, Textarea, Video, View, VoipRoom, WebView };
